lib/telemetry_sampler.ex

```diff
@@ -1,5 +1,400 @@
 defmodule Telemetry.Sampler do
   @moduledoc """
-  Allows to periodically gather measurements and publish them as `Telemetry` events
+  Allows to periodically collect measurements and publish them as `Telemetry` events.
+
+  ## Measurement specifications
+
+  Measurement specifications (later "specs") describe how the measurement should be performed and
+  what event should be emitted once the mesured value is collected. There are three possible values
+  for measurement spec:
+
+  1. An MFA tuple returning a single sample or a list of samples. Sample is an opaque data structure
+     which consists of event name, collected measurement value and event metadata. Samples can
+     be constructed using the `sample/3` function. Sampler will dispatch a `Telemetry` event
+     for each sample returned by the MFA invokation.
+  2. An atom representing a name of the function from `Telemetry.Sampler.VM` module. See documentation
+     for `Telemetry.Sampler.VM` module to learn about allowed values. Atom measurement spec is
+     equivalent to `{Telemetry.Sampler.VM, atom, []}` MFA spec.
```

Comment:
I was probably unclear when we talked about this but I think we should keep all VM samplers in a separate module decoupled from the main Telemetry.Sampler.

So basically, we would have `Telemetry.VMSampler` with one API, likely the `:atom` one, you describe here, and the `Telemetry.Sampler` only contains the general API.

Reply:
It would also be nice if we provide a complete example of how to measure the message queue of certain processes, as that will be the most common UI/usage.

Reply:
Hm, do you think it's valuable to track message queue lengths of certain processes? I've never done this, to be honest, because I don't know which processes to choose for tracking. Could you give me some example?

When it comes to message queues, I find the sum of message queues of all processes valuable, because it shows that something is not keeping up with the work in the VM, and requires further investigation (e.g. with the tool like `recon`).

Reply:
So a `VMSampler` would be just a Sampler instance with a default set of measurements for the VM? But anyone could just take the regular Sampler and build a VMSampler themselves using the functions from `Telemetry.Sampler.VM` module, is that what you mean?

Reply:
Thanks! Now I need to put these into the documentation 😄 

I agree re run queues, I have the implementation for tracking their length in rumbl-example repo.

Reply:
> So basically, we would have Telemetry.VMSampler with one API, likely the :atom one, you describe here, and the Telemetry.Sampler only contains the general API.

Wouldn't it be more clear to provide a function like `Telemetry.Sampler.VM.make_specs/1` which would translate the atom specs to the regular ones, including some kind of defaults? I'm afraid that providing the whole other module which only does this (translates atom specs to MFA) might introduce unnecessary confusion.

Reply:
Good point. What about `Telemetry.Sampler.vm_measurements`? I would like to avoid using the word specs because it is already overloaded in Elixir (supervisor specs and type specs).

Reply:
Great! I'll try to get rid of "spec"s in code, too 🙂